packages/trace-viewer/src/ui/llm.tsx

```diff
@@ -0,0 +1,259 @@
+
+    for await (const sseEvent of parseSSE(response.body)) {
+      const event = JSON.parse(sseEvent.data);
+      if (event.type  === 'content_block_delta')
+        yield event.delta.text;
+    }
+  }
+}
+
+class LLMChat {
+  conversations = new Map<string, Conversation>();
+
+  constructor(readonly api: LLM) {}
+
+  getConversation(id: string, systemPrompt: string) {
+    if (!this.conversations.has(id)) {
+      const conversation = new Conversation(this, systemPrompt);
+      this.conversations.set(id, conversation); // TODO: cleanup
+    }
+    return this.conversations.get(id)!;
+  }
+}
+
+export class Conversation {
+  history: LLMMessage[];
+  onChange = new EventEmitter<void>();
+  private _abortControllers = new Set<AbortController>();
+
+  constructor(public chat: LLMChat, systemPrompt: string) {
+    this.history = [{ role: 'developer', content: systemPrompt }];
+  }
+
+  async send(content: string, displayContent?: string) {
+    const response: LLMMessage = { role: 'assistant', content: '' };
+    this.history.push({ role: 'user', content, displayContent }, response);
+    const abortController = new AbortController();
+    this._abortControllers.add(abortController);
+    this.onChange.fire();
+    try {
+      for await (const chunk of this.chat.api.chatCompletion(this.history, abortController.signal)) {
+        response.content += chunk;
+        this.onChange.fire();
+      }
+    } finally {
+      this._abortControllers.delete(abortController);
+      this.onChange.fire();
+    }
+  }
+
+  isSending(): boolean {
+    return this._abortControllers.size > 0;
+  }
+
+  abortSending() {
+    for (const controller of this._abortControllers)
+      controller.abort();
+    this._abortControllers.clear();
+    this.onChange.fire();
+  }
+
+  isEmpty() {
+    return this.history.length < 2;
+  }
+}
+
+
+const llmContext = React.createContext<LLMChat | undefined>(undefined);
+
+export function LLMProvider({ children }: React.PropsWithChildren<{}>) {
+  const cookiePairs = useCookies();
+  const chat = React.useMemo(() => {
+    const cookies = Object.fromEntries(cookiePairs);
+    if (cookies.openai_api_key)
+      return new LLMChat(new OpenAI(cookies.openai_api_key, cookies.openai_base_url));
+    if (cookies.anthropic_api_key)
+      return new LLMChat(new Anthropic(cookies.anthropic_api_key, cookies.anthropic_base_url));
+  }, [cookiePairs]);
+  return <llmContext.Provider value={chat}>{children}</llmContext.Provider>;
+}
+
+export function useLLMChat() {
+  return React.useContext(llmContext);
+}
+
+export function useLLMConversation(id: string, systemPrompt: string) {
+  const chat = useLLMChat();
+  if (!chat)
+    throw new Error('No LLM chat available, make sure theres a LLMProvider above');
+  const conversation = React.useMemo(() => chat.getConversation(id, systemPrompt), [chat, id]); // eslint-disable-line react-hooks/exhaustive-deps
```

Comment:
In my experience, linter was always correct and I was always incorrect, so I always doubt disabling this rule 😄

Reply:
turns out that a little refactoring made it disappear :D